core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,253 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import java.util.concurrent.TimeUnit
+
+import scala.collection.mutable
+
+import org.apache.spark.SparkConf
+import org.apache.spark.Success
+import org.apache.spark.TaskEndReason
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.util.SystemClock
+import org.apache.spark.util.ThreadUtils
+import org.apache.spark.util.Utils
+import org.apache.spark.util.Clock
+
+/**
+ * BlacklistTracker is design to track problematic executors and node on application level.
+ * It is shared by all TaskSet, so that once a new TaskSet coming, it could be benefit from
+ * previous experience of other TaskSet.
+ *
+ * Once task finished, the callback method in TaskSetManager should update
+ * executorIdToFailureStatus Map.
+ */
+private[spark] class BlacklistTracker(
+    sparkConf: SparkConf,
+    clock: Clock = new SystemClock()) extends BlacklistCache{
+  // maintain a ExecutorId --> FailureStatus HashMap
+  private val executorIdToFailureStatus: mutable.HashMap[String, FailureStatus] = mutable.HashMap()
+
+  // Apply Strategy pattern here to change different blacklist detection logic
+  private val strategy = BlacklistStrategy(sparkConf)
+
+  // A daemon thread to expire blacklist executor periodically
+  private val scheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "spark-scheduler-blacklist-expire-timer")
+
+  private val recoverPeriod = sparkConf.getTimeAsSeconds(
+    "spark.scheduler.blacklist.recoverPeriod", "60s")
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireExecutorsInBlackList())
+      }
+    }
+    scheduler.scheduleAtFixedRate(scheduleTask, 0L, recoverPeriod, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  /** VisibleForTesting */
+  private[scheduler] def expireExecutorsInBlackList(): Unit = synchronized {
+    val updated = strategy.expireExecutorsInBlackList(executorIdToFailureStatus, clock)
+    if (updated) {
+      invalidateCache()
+    }
+  }
```

Comment:
@kayousterhout Would be good to have your thoughts here since I am fairly out of touch with scheduler now.

If I read it correctly, multiple methods in this class are directly called from scheduler as part of identifying, starting, updating tasks to be run from TaskSetManager/scheduler and they all use the same lock as updateFailedExecutors - which can actually run fairly non trivial computation as number of failures increase in the system. Is this a concern or is it alleviated somehow and I missed it ?